plugins/single_plugins/idle.cpp

```diff
@@ -4,69 +4,250 @@ extern "C"
 }
 
 #include "singleton-plugin.hpp"
+#include "render-manager.hpp"
 #include "output.hpp"
 #include "core.hpp"
 #include "config.hpp"
 #include "output-layout.hpp"
+#include "../cube/cube-control-signal.hpp"
+
+#include <animation.hpp>
+
+#define ZOOM_BASE 1.0
 
 class wayfire_idle
 {
+    wf::effect_hook_t render_hook;
+    double rotation = 0.0;
+    wf_duration duration;
+    wf_transition rot_transition, zoom_transition;
+    bool screensaver_stopping = false;
     bool idle_enabled = true;
-    wlr_idle_timeout *timeout = NULL;
-    wf::wl_listener_wrapper on_idle, on_resume;
+    bool screensaver_enabled = false;
+    bool outputs_inhibited = false;
+    wlr_idle_timeout *timeout_screensaver = NULL;
+    wlr_idle_timeout *timeout_dpms = NULL;
+    wf::wl_listener_wrapper on_idle_screensaver, on_resume_screensaver;
+    wf::wl_listener_wrapper on_idle_dpms, on_resume_dpms;
 
-    wf_option dpms_timeout;
+    wf_option dpms_timeout, screensaver_timeout;
+    wf_option cube_zoom_speed, cube_rotate_speed, cube_zoom_end;
     wf_option_callback dpms_timeout_updated = [=] () {
-        create_timeout(dpms_timeout->as_int());
+        create_dpms_timeout(dpms_timeout->as_int());
+    };
+    wf_option_callback screensaver_timeout_updated = [=] () {
+        create_screensaver_timeout(screensaver_timeout->as_int());
+    };
+    wf_option_callback cube_zoom_speed_updated = [=] () {
```

Comment:
You don't need this, if you create a wf_duration from a wf_option, it will automatically update when the option updates.